functions/src/webhooks/transcription.ts

```diff
@@ -1,57 +1,102 @@
-              }
-              return false
+        if (maybeEventsInDb.docs.length) {
+          // If we have a match look for one that matches a hash of the token
+          // we gave Assembly. There should only be one of these but firestore
+          // gives us an array. If there is more than one member, something is
+          // wrong
+          const authenticatedEventIds = [] as string[]
+          const hashedToken = sha256(String(req.headers["x-maple-webhook"]))
+
+          for (const index in maybeEventsInDb.docs){
+            const doc = maybeEventsInDb.docs[index]
+
+            const tokenDocInDb = await db
+              .collection("events")
+              .doc(doc.id)
+              .collection("private")
+              .doc("webhookAuth")
+              .get()
+
+            const tokenDataInDb = tokenDocInDb.data()?.videoAssemblyWebhookToken
+
+            if (hashedToken === tokenDataInDb) {
+              authenticatedEventIds.push(doc.id)
             }
-          )
-          if (authenticatedEventsInDb) {
+          }
+
+          if (authenticatedEventIds.length === 1) {
+            // If there is one authenticated event, pull out the parts we want to
+            // save and try to save them in the db.
+            const { id, text, audio_url, utterances } = transcript
             try {
-              await db
+              const transcriptionInDb = await db
                 .collection("transcriptions")
-                .doc(transcript.id)
-                .set({ _timestamp: new Date(), ...transcript })
+                .doc(id)
 
-              authenticatedEventsInDb.forEach(async d => {
-                await d.ref.update({
-                  ["webhook_auth_header_value"]: null
+              await transcriptionInDb.set({
+                id,
+                text,
+                createdAt: Timestamp.now(),
+                audio_url
+              })
+
+              // Put each `utterance` in a separate doc in an utterances
+              // collection. Previously had done the same for `words` but
+              // got worried about collection size and write times since
+              // `words` can be tens of thousands of members.
+              if (utterances) {
+                const writer = db.bulkWriter()
+                for (let utterance of utterances) {
+                  const { speaker, confidence, start, end, text } = utterance
+
+                  writer.set(
+                    db
+                      .collection("transcriptions")
+                      .doc(`${transcript.id}`)
+                      .collection("utterances")
+                      .doc(),
+                    { speaker, confidence, start, end, text }
+                  )
+                }
+
+                await writer.close()
+              }
+
+              // Delete the hashed webhook auth token from our db now that
+              // we're done.
+              authenticatedEventIds.forEach(async docId => {
```

Comment:
nit: I think this `async forEach` could run into the same problem as the other one that was just fixed - there's no guarantee the async function will complete before the request ends.